torch/utils/cpp_extension.py

```diff
@@ -0,0 +1,188 @@
+            to the build.
+        build_directory: optional path to use as build workspace.
+        verbose: If `True`, turns on verbose logging of load steps.
+
+    Returns:
+        The loaded PyTorch extension as a Python module.
+    '''
+
+    # Allows sources to be a single path or a list of paths.
+    if isinstance(sources, str):
+        sources = [sources]
+
+    if build_directory is None:
+        build_directory = _get_build_directory(name, verbose)
+
+    build_file_path = os.path.join(build_directory, 'build.ninja')
+    if verbose:
+        print('Emitting ninja build file {}...'.format(build_file_path))
+    # NOTE: Emitting a new ninja build file does not cause re-compilation if
+    # the sources did not change, so it's ok to re-emit (and it's fast).
+    _write_ninja_file(
+        build_file_path,
+        name,
+        sources,
+        extra_cflags or [],
+        extra_ldflags or [],
+        extra_include_paths or [])
+
+    if verbose:
+        print('Building extension module {}...'.format(name))
+    _build_extension_module(name, build_directory)
+
+    if verbose:
+        print('Loading extension module {}...'.format(name))
+    return _import_module_from_library(name, build_directory)
+
+
+def _get_build_directory(name, verbose):
+    root_extensions_directory = os.environ.get('TORCH_EXTENSIONS_DIR')
+    if root_extensions_directory is None:
+        # tempfile.gettempdir() will be /tmp on UNIX and \TEMP on Windows.
+        root_extensions_directory = os.path.join(tempfile.gettempdir(),
+                                                 'torch_extensions')
+
+    if verbose:
+        print('Using {} as PyTorch extensions root...'.format(
+            root_extensions_directory))
+
+    build_directory = os.path.join(root_extensions_directory, name)
+    if not os.path.exists(build_directory):
+        if verbose:
+            print('Creating extension directory {}...'.format(build_directory))
+        # This is like mkdir -p, i.e. will also create parent directories.
+        os.makedirs(build_directory)
+
+    return build_directory
+
+
+def _build_extension_module(name, build_directory):
+    try:
+        subprocess.check_output(
+            ['ninja', '-v'], stderr=subprocess.STDOUT, cwd=build_directory)
+    except subprocess.CalledProcessError:
+        # Python 2 and 3 compatible way of getting the error object.
+        _, error, _ = sys.exc_info()
+        # error.output contains the stdout and stderr of the build attempt.
+        raise RuntimeError("Error building extension '{}': {}".format(
+            name, error.output.decode()))
+
+
+def _import_module_from_library(module_name, path):
+    # https://stackoverflow.com/questions/67631/how-to-import-a-module-given-the-full-path
+    file, path, description = imp.find_module(module_name, [path])
+    # Close the .so file after load.
+    with file:
+        return imp.load_module(module_name, file, path, description)
```

Comment:
does this work in Python 2 and Python 3?

Reply:
Yep I just tried in two different conda/virtual environments with Python 2.7 and 3.6